### PR TITLE
parallel_for control flow ops - change for_loop output to match pfor when run with 0 iters

### LINKED ISSUE #37795: GradientTape.jacobian works for batch shape 0 when `experimental_use_pfor=True` but not when `experimental_use_pfor=False`
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow):  No
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04):  `macOS Catalina 10.15.2 (19C57)`
- TensorFlow installed from (source or
binary): - TensorFlow version (use command below): `v2.2.0-rc0-43-gacf4951a2f 2.2.0-rc1 0.10.0-dev20200321`
- Python version: 3.7.6
- CUDA/cuDNN version: - GPU model and memory: n/a

**Describe the current behavior**
Using `experimental_use_pfor=False` in `GradientTape.jacobian` behaves otherwise similarly as using `experimental_use_pfor=True`, but when running with tensors of batch shape 0, the former fails due to a type error, whereas the latter one works fine.

**Describe the expected behavior**
I would expect the behavior to be the same in both cases when `experimental_use_pfor=False` and `experimental_use_pfor=True`.

**Standalone code to reproduce the issue**
```python
import tensorflow as tf


def main():
    variable = tf.Variable(1.0)
    inputs = (
        tf.constant(tf.random.uniform((0, 4))),
        tf.constant(tf.random.uniform((0, 3))),
    )

    with tf.GradientTape(persistent=True) as tape:
        outputs = variable * tf.pow(tf.concat(inputs, axis=-1), 2.0)

    jacobians_1 = tape.jacobian(
        outputs,
        variable,
        experimental_use_pfor=True,
    )
    print(jacobians_1)
    print("tape.jacobians(..., experimental_use_pfor=True) works!")

    try:
        jacobians_2 = tape.jacobian(
            outputs,
            variable,
            experimental_use_pfor=False,
        )
        print(jacobians_2)
        print("tape.jacobians(..., experimental_use_pfor=False) works!")
    except TypeError:
        print("tape.jacobians(..., experimental_use_pfor=False) doesn't work!")
        raise


if __name__ == '__main__':
    main()
```

**Other info / logs**
Originally posted here: https://github.com/tensorflow/tensorflow/issues/32460#issuecomment-572545864.

```
tape.jacobians(..., experimental_use_pfor=True) works!
tape.jacobians(..., experimental_use_pfor=False) doesn't work!
Traceback (most recent call last):
  File "./tests/test_tape_jacobian.py", line 36, in <module>
    main()
  File "./tests/test_tape_jacobian.py", line 26, in main
    experimental_use_pfor=False,
  File "/Users/hartikainen/conda/envs/softlearning-tf2/lib/python3.7/site-packages/tensorflow/python/eager/backprop.py", line 1151, in jacobian
    for i, out in enumerate(output):
TypeError: 'NoneType' object is not iterable
```

### PR DESCRIPTION
**Pull request by jjhbriggs**
When run with zero iterations, tensorflow.python.ops.parallel_for.control_flow_ops.pfor returns a tensor with a shape of 0 in the first dimension (with the corresponding return shape in the others). However, running tensorflow.python.ops.parallel_for.control_flow_ops.for_loop on the same inputs with zero iterations would return `None`.

This manifests itself in issues such as #37795, where using experimental_use_pfor=True for computing Jacobians (with GradientTape) on inputs with dimension 0 would 'work', and return a tensor with shape 0, however running with experimental_use_pfor=False would throw a Type Error. 

This PR changes the for_loop handling to return an object identical to that of pfor when run with zero iterations, rather than `None`.

--
Resolves #37795.

### CODE REVIEW

**Review by tomytsai1 — APPROVED** [review]
Thanks for the fix!

**Review by tomytsai1 — CHANGES_REQUESTED** [review]
Some unit tests failed ([log](https://source.cloud.google.com/results/invocations/16ea0468-1a35-4a19-8697-0df724733551/targets/tensorflow%2Frel%2Fdocker%2Fgithub_presubmit%2Fpycpp_cpu_py39/log)): from `tensorflow/python/eager/backprop_test.py`
- BatchJacobianTest.test_degenerate_shape3
- JacobianTest.test_empty_tensor_consistent_jacobian
Please fix them.

**Comment by jjhbriggs** [comment]
Apologies, these should be resolved now.

**Comment by tomytsai1** [comment]
The tests "Code Check - Changed Files" and "Py+CPP Test Suite - Ubuntu CPU, Python 3.9" seemed to be stuck... Will see how to restart them for validating the change.

File: tensorflow/python/ops/parallel_for/control_flow_ops.py [hunk 1]
```diff
@@ -104,9 +105,18 @@ def while_body(i, *ta_list):
...
+    try:
+      loop_fn_out = loop_fn(loop_var)
+    except InvalidArgumentError: 
+      loop_fn_out = array_ops.placeholder_with_default(0, shape=[])
+    out_shapes = [[0]+ops.convert_to_tensor(x).shape
+            for x in nest.flatten(loop_fn_out)]
+    output = [array_ops.zeros(out_shapes[i], dt)
+            for i,dt in enumerate(flat_loop_fn_dtypes)]
+
+  return nest.pack_sequence_as(loop_fn_dtypes, output)
```
Comment (tomytsai1) [inline tensorflow/python/ops/parallel_for/control_flow_ops.py:119]: Thanks for the fix, and we are running the checks to see if there are other tests that also need attention.

Also am wondering
- Could we catch more/all exceptions in addition to `InvalidArgumentError` as there is not clear contract about which exception type should be thrown for  `loop_fn` when the input is empty ?

- Is it possible that the, in the exception case,  line 116-117 would fail for `len(flat_loop_fn_dtypes) != out_shapes` as we computed `out_shapes` based on the `loop_fn_out` of empty shape placeholder.   In this case, could we directly use a logic to compute the output as an empty tensor with no shape in the `except` block ?

Comment (jjhbriggs) [reply tensorflow/python/ops/parallel_for/control_flow_ops.py]: 1. Yes that makes sense, changed to catch all Exceptions.
2. I'm not sure, but directly computing the output appears clearer anyway, so I have made that change as well

**Comment by tomytsai1** [comment]
Thanks for the fix! Triggering the tests again to see if there is any undetected issues.

**Comment by tomytsai1** [comment]
The [ARM CI build](https://github.com/tensorflow/tensorflow/actions/runs/4189182411/jobs/7261197447) failed in many unit tests with similar stack traces
```
...
2023-02-16T21:38:51.9389644Z   File "/workspace/pip_test/venv_clean/lib/python3.10/site-packages/tensorflow/python/framework/ops.py", line 3627, in _create_op_internal
2023-02-16T21:38:51.9389809Z     ret = Operation.from_node_def(
2023-02-16T21:38:51.9390303Z   File "/workspace/pip_test/venv_clean/lib/python3.10/site-packages/tensorflow/python/framework/ops.py", line 2044, in from_node_def
2023-02-16T21:38:51.9390542Z     c_op = _create_c_op(g, node_def, inputs, control_input_ops, op_def=op_def)
2023-02-16T21:38:51.9391043Z   File "/workspace/pip_test/venv_clean/lib/python3.10/site-packages/tensorflow/python/util/traceback_utils.py", line 141, in error_handler
2023-02-16T21:38:51.9391192Z     return fn(*args, **kwargs)
2023-02-16T21:38:51.9391680Z   File "/workspace/pip_test/venv_clean/lib/python3.10/site-packages/tensorflow/python/framework/ops.py", line 1870, in _create_c_op
2023-02-16T21:38:51.9392051Z     op_def = graph.op_def_for_type(node_def.op)  # pylint: disable=protected-access
2023-02-16T21:38:51.9392561Z   File "/workspace/pip_test/venv_clean/lib/python3.10/site-packages/tensorflow/python/framework/ops.py", line 4031, in op_def_for_type
2023-02-16T21:38:51.9392782Z     self._op_def_cache[type] = op_def_pb2.OpDef.FromString(
```
Triggering the build again to see if this is a transient error.

**Review by tomytsai1 — APPROVED** [review]
LGTM. Thanks for supporting this corner case.

The two tests of ARM CI / build (3.10) (pull_request) showed different types of errors. Possibly they are related to test infra but not this change.
